### PR TITLE
feat(docs): add search capability

### PR DESCRIPTION
**Pull request by lucivpav**
This PR adds a search input to the top part of sidebar. The search goes through all the below tree's sections, effectively filtering those that match the query. The items are matched in case-insensitive fashion.

search

As you can see, the icon is not rendered at an expected position. This looks like a bug in Input component, reported [here](https://github.com/stardust-ui/react/issues/1678).

The search input is not fully accessible - it does not let the user know, how many items were found (matching the query). This would be solved in a future PR.

You can search by typing even when the element is not focused (provided no other input element is in focus).

### CODE REVIEW

**Comment by lucivpav** [comment]
Problem: category of currently selected section cannot be collapsed anymore.

**Comment by lucivpav** [comment]
> Problem: category of currently selected section cannot be collapsed anymore.

Fixed.

**Comment by lucivpav** [comment]
Icon's location has been fixed by manually setting the icon's padding.

image

File: docs/src/components/Sidebar/Sidebar.tsx [hunk 1]
```diff
@@ -479,10 +434,41 @@ class Sidebar extends React.Component<any, any> {
...
+            </Box>
+          </a>
+          <a href={changeLogUrl} target="_blank" rel="noopener noreferrer" style={topItemTheme}>
+            <Box>
+              CHANGELOG
+              <Icon name="file alternate outline" styles={{ float: 'right' }} />
+            </Box>
+          </a>
+          <Input
+            style={topItemTheme}
```
Comment (mnajdova) [inline docs/src/components/Sidebar/Sidebar.tsx]: Please always use `styles` instead of `style` with Stardust components. Also, instead of padding, try applying margin and everything should work. The icon inside the menu is positioned absolutelly, that's why padding is not working: https://codesandbox.io/s/stardust-ui-example-i4e6k

File: docs/src/components/Sidebar/Sidebar.tsx [hunk 2]
```diff
@@ -38,6 +39,17 @@ class Sidebar extends React.Component<any, any> {
...
   selectedRoute: any
   filteredMenu = componentMenu
+  treeRef: Ref
+  searchInputRef: Ref
+
+  constructor(props) {
+    super(props)
+    this.handleQueryChange = this.handleQueryChange.bind(this)
+    this.findActiveCategoryIndex = this.findActiveCategoryIndex.bind(this)
+    this.setSearchInput = this.setSearchInput.bind(this)
```
Comment (mnajdova) [inline docs/src/components/Sidebar/Sidebar.tsx]: Please use arrow functions instead of binding this methods in the constructor.

**Comment by mnajdova** [comment]
I would expect that Concepts, Guides and Prototypes would not be visible in this case
image

File: docs/src/components/Sidebar/Sidebar.tsx (hunk 1 above)
Comment (lucivpav) [reply docs/src/components/Sidebar/Sidebar.tsx]: I think that in this case, the code will look cleaner if I stick to padding.

**Comment by lucivpav** [comment]
> I would expect that Concepts, Guides and Prototypes would not be visible in this case
> image

Good point, fixed.

File: docs/src/components/Sidebar/Sidebar.tsx [hunk 3]
```diff
@@ -488,10 +445,41 @@ class Sidebar extends React.Component<any, any> {
...
+              <Icon name="github" styles={{ float: 'right' }} />
+            </Box>
+          </a>
+          <a href={changeLogUrl} target="_blank" rel="noopener noreferrer" style={topItemTheme}>
+            <Box>
+              CHANGELOG
+              <Icon name="file alternate outline" styles={{ float: 'right' }} />
+            </Box>
+          </a>
+          <Input
```
Comment (mnajdova) [inline docs/src/components/Sidebar/Sidebar.tsx:457]: I think it would look better if we use the clearable input with the search icon positioned in the end. What do you think @lucivpav

File: docs/src/components/Sidebar/Sidebar.tsx [hunk 4]
```diff
@@ -38,6 +39,14 @@ class Sidebar extends React.Component<any, any> {
   _searchInput: any
   selectedRoute: any
   filteredMenu = componentMenu
+  treeRef: Ref
+  searchInputRef: Ref
```
Comment (layershifter) [inline docs/src/components/Sidebar/Sidebar.tsx]: Oops, it should be

```suggestion
  searchInputRef: React.RefObject<HTMLElement>
```

However, you don't need `constructor()` 🐱 
```suggestion
  searchInputRef: React.createRef()
```

File: docs/src/components/Sidebar/Sidebar.tsx [hunk 5]
```diff
@@ -488,10 +445,41 @@ class Sidebar extends React.Component<any, any> {
...
+          <Input
+            styles={topItemTheme}
+            fluid
+            icon={{ name: 'search', style: { padding: topItemTheme.padding } }}
+            placeholder="Search"
+            iconPosition="start"
+            role="search"
+            onChange={this.handleQueryChange}
+            value={this.state.query}
+            ref={this.searchInputRef}
```
Comment (layershifter) [inline docs/src/components/Sidebar/Sidebar.tsx]:
```suggestion
            inputRef={this.searchInputRef}
```

We have `inputRef`, so we can simplify it

File: docs/src/components/Sidebar/Sidebar.tsx [hunk 6]
```diff
@@ -52,9 +61,14 @@ class Sidebar extends React.Component<any, any> {
     document.removeEventListener('keydown', this.handleDocumentKeyDown)
   }
 
-  setSearchInput() {
-    // TODO: Replace findDOMNode with Ref component when it will be merged
-    this._searchInput = (findDOMNode(this) as any).querySelector('.ui.input input')
+  setSearchInput = () => {
+    this._searchInput = this.searchInputRef.current.inputRef.current
+  }
```
Comment (layershifter) [inline docs/src/components/Sidebar/Sidebar.tsx]: You don't need this, check this one: https://github.com/stardust-ui/react/pull/1682/files#r307795621

File: docs/src/components/Sidebar/Sidebar.tsx [hunk 7]
```diff
@@ -66,67 +80,17 @@ class Sidebar extends React.Component<any, any> {
     if (!hasModifier && isAZ && bodyHasFocus) this._searchInput.focus()
```
Comment (layershifter) [inline docs/src/components/Sidebar/Sidebar.tsx]:
```suggestion
    if (!hasModifier && isAZ && bodyHasFocus) this._searchInput.current.focus()
```

https://github.com/stardust-ui/react/pull/1682/files#r307795621

File: docs/src/components/Sidebar/Sidebar.tsx [hunk 8]
```diff
@@ -63,69 +64,23 @@ class Sidebar extends React.Component<any, any> {
...
-          as={NavLink}
-          to={getComponentPathname(info)}
-        >
-          {info.displayName}
-          {isSelected && selectedItemLabel}
-        </Menu.Item>
-      )
-    })
-*/
+  treeActiveIndexChanged = data => {
```
Comment (mnajdova) [inline docs/src/components/Sidebar/Sidebar.tsx]: Usually these kind of callbacks sends two params here:
- event
- all props

So this one should be refactored to something like:
```
treeActiveIndexChange = (e, props) => {
 this.setState({activeCategoryIndex: props.activeIndex})
}
```

Comment (mnajdova) [reply docs/src/components/Sidebar/Sidebar.tsx]: Take a look on the `onMenuOpenChange` in the `MenuItem` component.

File: docs/src/components/Sidebar/Sidebar.tsx [hunk 9]
```diff
@@ -488,10 +433,42 @@ class Sidebar extends React.Component<any, any> {
...
+            value={this.state.query}
+            inputRef={this.searchInputRef}
+          />
+        </Flex>
         <Tree
-          defaultActiveIndex={activeCategoryIndex}
           items={allSections}
           renderItemTitle={titleRenderer}
+          activeIndex={this.state.activeCategoryIndex}
+          onActiveIndexChange={this.treeActiveIndexChanged}
```
Comment (mnajdova) [inline docs/src/components/Sidebar/Sidebar.tsx:474]: Let's add the method in the Tree in this PR, so we can test it with the real implementation. Then we can extract it to a separate PR

Comment (lucivpav) [reply docs/src/components/Sidebar/Sidebar.tsx:474]: Ok. Please take a look: https://github.com/stardust-ui/react/pull/1728

File: packages/react/src/components/Tree/Tree.tsx [hunk 10]
```diff
@@ -52,6 +53,13 @@ export interface TreeProps extends UIComponentProps, ChildrenComponentProps {
    * @param {ReactNode|ReactNodeArray} children - The computed children for this slot.
    */
   renderItemTitle?: ShorthandRenderFunction
+
+  /** Called when activeIndex changes.
+   *
+   * @param {SyntheticEvent} event - React's original SyntheticEvent.
+   * @param {object} data - All props and proposed value.
+   */
+  onActiveIndexChange?: ComponentEventHandler<TreeProps & { activeIndex: number[] | number }>
```
Comment (mnajdova) [inline packages/react/src/components/Tree/Tree.tsx]:
```suggestion
  onActiveIndexChange?: ComponentEventHandler<TreeProps>
```
The `TreeProps` interface already contains the `activeIndex`

File: docs/src/components/Sidebar/Sidebar.tsx (hunk 3 above)
Comment (lucivpav) [reply docs/src/components/Sidebar/Sidebar.tsx:457]: I moved the icon to the right side, because if we have the input `clearable`, it looks bad on the left side.

File: docs/src/components/Sidebar/Sidebar.tsx [hunk 11]
```diff
@@ -288,79 +244,24 @@ class Sidebar extends React.Component<any, any> {
...
-      fontWeight: 700,
-    }
-
-    const flexDisplayStyle: any = { width: `${parseInt(this.props.width, 10) - 50}px` }
-
-    const changeLogUrl: string = `${constants.repoURL}/blob/master/CHANGELOG.md`
+  handleQueryChange = e => {
+    const query = e.target.value ? e.target.value : ''
+    this.setState({ query })
+  }
```
Comment (layershifter) [inline docs/src/components/Sidebar/Sidebar.tsx:247]:
```jsx
handleQueryChange = (e, data) => {
  this.setState({ query: data.value })
}
```

We have `value` in the second param on `Input`

File: docs/src/components/Sidebar/Sidebar.tsx [hunk 12]
```diff
@@ -34,27 +34,28 @@ class Sidebar extends React.Component<any, any> {
     history: PropTypes.object.isRequired,
     style: PropTypes.object,
   }
-  state: any = { query: '' }
+  state: any = { query: '', activeCategoryIndex: 0 }
   _searchInput: any
   selectedRoute: any
   filteredMenu = componentMenu
+  searchInputRef: React.RefObject<HTMLInputElement> = React.createRef()
```
Comment (layershifter) [inline docs/src/components/Sidebar/Sidebar.tsx]:
```suggestion
  searchInputRef = React.createRef<HTMLInputElement>()
```

nit:

File: docs/src/components/Sidebar/Sidebar.tsx [hunk 13]
```diff
@@ -34,27 +34,28 @@ class Sidebar extends React.Component<any, any> {
...
 
   componentDidMount() {
     document.addEventListener('keydown', this.handleDocumentKeyDown)
-    this.setSearchInput()
-  }
 
-  componentDidUpdate() {
-    this.setSearchInput()
+    const at = window.location.href.replace(window.location.origin, '')
```
Comment (layershifter) [inline docs/src/components/Sidebar/Sidebar.tsx]: Can we use `this.props.location.pathname`?

`Sidebar` is wrapped with `withRouter()`:
https://github.com/ReactTraining/react-router/blob/master/packages/react-router/docs/api/withRouter.md

File: docs/src/components/Sidebar/Sidebar.tsx [hunk 14]
```diff
@@ -63,69 +64,23 @@ class Sidebar extends React.Component<any, any> {
...
-  handleItemClick = () => {
+  handleItemClick = e => {
     const { query } = this.state
 
-    if (query) this.setState({ query: '' })
-    // TODO: as part of search input re-enabling
-    // if (document.activeElement === this._searchInput) this._searchInput.blur()
+    if (query) {
+      this.setState({ query: '' })
+      const at = e.target.href.replace(e.target.baseURI, '/')
```
Comment (layershifter) [inline docs/src/components/Sidebar/Sidebar.tsx]: We can get value from `data` somehow?

```tsx
handleItemClick = (e: React.SyntheticEvent, data: TreeItemProps) => {
  const at = data.title.to // ???
}
```

File: docs/src/components/Sidebar/Sidebar.tsx [hunk 15]
```diff
@@ -63,69 +64,23 @@ class Sidebar extends React.Component<any, any> {
...
-          as={NavLink}
-          to={getComponentPathname(info)}
-        >
-          {info.displayName}
-          {isSelected && selectedItemLabel}
-        </Menu.Item>
-      )
-    })
-*/
+  treeActiveIndexChanged = (e, props) => {
```
Comment (layershifter) [inline docs/src/components/Sidebar/Sidebar.tsx]:
```suggestion
  treeActiveIndexChanged = (e: React.SyntheticEvent, props: TreeItemProps) => {
```

Can we use types?

File: docs/src/components/Sidebar/Sidebar.tsx [hunk 16]
```diff
@@ -449,31 +350,78 @@ class Sidebar extends React.Component<any, any> {
...
+    const allSections = _.filter(
+      allSectionsWithPossibleEmptySections,
+      (section: ShorthandValue<TreeItemProps>) => (section as any).items.length > 0,
+    )
+
+    if (this.state.query !== '') {
+      // open all sections
+      _.forEach(allSections, (section: ShorthandValue<TreeItemProps>) => {
+        ;(section as any).open = true
+      })
```
Comment (layershifter) [inline docs/src/components/Sidebar/Sidebar.tsx:400]: Can we avoid mutation?

```tsx
allSections = _.map(allSections, (section: ShorthandValue<TreeItemProps>) => {
  return { ...section, open: true }
})
```

File: docs/src/components/Sidebar/Sidebar.tsx [hunk 17]
```diff
@@ -449,31 +350,78 @@ class Sidebar extends React.Component<any, any> {
...
+    const changeLogUrl: string = `${constants.repoURL}/blob/master/CHANGELOG.md`
+    const allSectionsWithoutSearchFilter = this.getSectionsWithoutSearchFilter()
+
+    const allSectionsWithPossibleEmptySections = _.map(
+      allSectionsWithoutSearchFilter,
       (section: ShorthandValue<TreeItemProps>) => {
-        return _.find((section as any).items, item => item.title.to === at)
+        if ((section as any).items) {
+          ;(section as any).items = _.filter((section as any).items, item =>
+            item.title.content.toLowerCase().includes(this.state.query.toLowerCase()),
```
Comment (layershifter) [inline docs/src/components/Sidebar/Sidebar.tsx]: Use regexp instead?

```tsx
const escapedQuery = _.escapeRegExp(this.state.query)
const regexQuery = new RegExp(`^${escapedQuery}`, 'i')
```

```tsx
(section: ShorthandValue<TreeItemProps>) => {
  return {
    ...section,
    items: _.filter(section.items, item => regexQuery.test(item.title)) // _.filter works with undefined
  }
}
```

**Review by layershifter — APPROVED** [review]
